client/starwhale/core/job/scheduler.py

```diff
@@ -0,0 +1,258 @@
+        elif query.kind == EvaluationResultKind.METRIC:
+            return self._datastore.get_metrics()
+
+    def __split_tasks(
+        self,
+        **kw: t.Any,
+    ) -> None:
+        for _step in self.steps:
+            # update step status = init
+            _step.status = STATUS.INIT
+            for index in range(_step.task_num):
+                _step.gen_task(
+                    index=index,
+                    module=self.module,
+                    workdir=self.workdir,
+                    src_dir=self.src_dir,
+                    dataset_uris=self.dataset_uris,
+                    version=self.version,
+                    project=self.project,
+                    **kw,
+                )
+
+    def schedule(self) -> None:
+        _threads: t.List[Executor] = []
+        with self._lock:
+            _wait_steps = []
+            _finished_step_names = []
+            for _step in self.steps:
+                if _step.status is STATUS.FAILED:
+                    # todo break processing
+                    pass
+                if _step.status is STATUS.SUCCESS:
+                    _finished_step_names.append(_step.step_name)
+                if _step.status is STATUS.INIT:
+                    _wait_steps.append(_step)
+            # judge whether a step's dependency all in finished
+            for _wait in _wait_steps:
+                if all(d in _finished_step_names for d in _wait.dependency if d):
+                    _wait.status = STATUS.RUNNING
+                    # add pipe handler
+
+                    for _t in _wait.tasks:
+                        _d = TaskDaemon(
+                            TaskPipe(
+                                idx=f"{_wait.step_name}-{_t.context.index}",
+                                main_conn=_t.main_conn,
+                                sub_conn=_t.sub_conn,
+                            ),
+                            self,
+                        )
+                        _d.start()
+                    _executor = Executor(
+                        _wait.concurrency, _wait, _wait.tasks, StepCallback(self)
+                    )
+                    _executor.start()
+                    # executor.setDaemon()
+                    _threads.append(_executor)
+
+        for _thread in _threads:
+            _thread.join()
+
+    def schedule_single_task(self, step_name: str, task_index: int) -> None:
+        _steps = [step for step in self.steps if step_name == step.step_name]
+        if len(_steps) == 0:
+            raise RuntimeError(f"step:{step_name} not found")
+        _step = _steps[0]
+        if task_index >= _step.task_num:
+            raise RuntimeError(
+                f"task_index:{task_index} out of bounds, total:{_step.task_num}"
+            )
+        _task = _step.tasks[task_index]
+        logger.debug(
+            "all result size: {}", len([res for res in self._datastore.get_results()])
```

Comment:
`len([res for res in self._datastore.get_results()])` -> `list(self._datastore.get_results())` ?

Reply:
`all results size` ?